File: app/src/main/java/io/github/qwqgong/androidcyaml/runtime/RuntimeCoordinator.java
package io.github.qwqgong.androidcyaml;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import org.json.JSONException;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/** Serializes every transition that can replace the TUN or embedded mihomo runtime. */
final class RuntimeCoordinator implements
        RuntimeConfigTransactions.Lifecycle,
        RuntimeConfigTransactions.Host,
        NetworkReconciler.Host {
    interface OperationCallback {
        void onComplete(boolean success, String detail);
    }

    private record RuntimeStartRequest(
            RuntimeOverrideSettings settings,
            String detailSuffix
    ) {
    }

    private static final String TAG = "AndroidCyaml/Coordinator";
    private static final int ANDROID_17_API = 37;
    private static volatile RuntimeCoordinator instance;

    private final Context context;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final RuntimeStateBus stateBus = new RuntimeStateBus();
    private final MihomoFileStore fileStore;
    private final ConfigInstaller configInstaller;
    private final RuntimeOverrideStore overrideStore;
    private final RuntimeConfigTransactions configTransactions;
    private final SelectorSession selectorSession;
    private final RuntimeLifecycle lifecycle;
    private final NetworkReconciler networkReconciler;

    private RuntimeCoordinator(Context context) {
        this.context = context.getApplicationContext();
        fileStore = new MihomoFileStore(this.context);
        configInstaller = new ConfigInstaller(this.context, fileStore);
        overrideStore = new RuntimeOverrideStore(this.context);
        Ipv6EnvironmentMonitor networkMonitor = new Ipv6EnvironmentMonitor(this.context);
        selectorSession = new SelectorSession(
                new NetworkSelectionStore(this.context),
                networkMonitor
        );
        lifecycle = new RuntimeLifecycle(fileStore, selectorSession);
        networkReconciler = new NetworkReconciler(
                networkMonitor,
                overrideStore,
                lifecycle,
                selectorSession,
                this
        );
        configTransactions = new RuntimeConfigTransactions(
                configInstaller,
                overrideStore,
                this,
                this
        );
    }

    static RuntimeCoordinator getInstance(Context context) {
        RuntimeCoordinator local = instance;
        if (local == null) {
            synchronized (RuntimeCoordinator.class) {
                local = instance;
                if (local == null) {
                    local = new RuntimeCoordinator(context);
                    instance = local;
                }
            }
        }
        return local;
    }

    static int trimMemoryCachesIfCreated() {
        RuntimeCoordinator local = instance;
        MihomoRuntime current = local == null ? null : local.lifecycle.runtime();
        // Do not initialize the large native core merely because Android asks
        // an otherwise idle service process to trim memory.
        return current == null ? 0 : current.trimRebuildableCaches();
    }

    static boolean persistStateForMemoryKill() {
        RuntimeCoordinator local = instance;
        if (local == null || local.lifecycle.runtime() == null) {
            return true;
        }
        Future<Boolean> checkpoint = local.executor.submit(
                () -> local.selectorSession.checkpoint(local.lifecycle.runtime())
        );
        try {
            return checkpoint.get(2L, TimeUnit.SECONDS);
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            return false;
        } catch (ExecutionException | TimeoutException exception) {
            Log.w(TAG, "Unable to persist selector choices before memory kill", exception);
            return false;
        }
    }

    RuntimeOverrideSettings runtimeOverrideSettings() {
        return overrideStore.settings();
    }

    @Override
    public boolean effectiveIpv6Enabled() {
        return lifecycle.effectiveIpv6Enabled();
    }

    void addListener(RuntimeStateBus.Listener listener) {
        stateBus.addListener(listener);
    }

    void removeListener(RuntimeStateBus.Listener listener) {
        stateBus.removeListener(listener);
    }

    void start(AndroidVpnService requestedService) {
        executor.execute(() -> startInternal(requestedService));
    }

    void stop(AndroidVpnService requestedService, Runnable completion) {
        executor.execute(() -> {
            if (lifecycle.service() != null
                    && requestedService != null
                    && !lifecycle.ownsService(requestedService)) {
                postCompletion(completion);
                return;
            }
            if (stateBus.snapshot().state() != RuntimeState.STOPPED) {
                publish(RuntimeState.STOPPING, "正在停止 mihomo JNI TUN…");
            }
            cleanupAll();
            publish(RuntimeState.STOPPED, "VPN 未连接");
            postCompletion(completion);
        });
    }

    void restart(OperationCallback callback) {
        executor.execute(() -> {
            if (!active()) {
                complete(callback, false, "VPN 未启动");
                return;
            }
            try {
                publish(RuntimeState.STARTING, "正在重启 mihomo JNI 核心…");
                String detail = start();
                publish(RuntimeState.RUNNING, detail);
                complete(callback, true, "mihomo 已重启");
            } catch (Exception exception) {
                String message = Exceptions.usefulMessage(exception);
                failActiveService("mihomo 重启失败：" + message);
                complete(callback, false, message);
            }
        });
    }

    void importConfig(Uri source, OperationCallback callback) {
        executor.execute(() -> configTransactions.importConfig(source, callback));
    }

    void setRuntimeOverrides(
            String processMatchingModeValue,
            boolean ipv6Enabled,
            String logLevelValue,
            boolean adaptiveTcpConcurrent,
            boolean webViewXhttp,
            boolean lanWebUiPublic,
            OperationCallback callback
    ) {
        final RuntimeLogLevel logLevel;
        final ProcessMatchingMode processMatchingMode;
        try {
            logLevel = RuntimeLogLevel.fromWireValue(logLevelValue);
            processMatchingMode = ProcessMatchingMode.fromWireValue(processMatchingModeValue);
        } catch (IllegalArgumentException exception) {
            complete(callback, false, Exceptions.usefulMessage(exception));
            return;
        }
        executor.execute(() -> configTransactions.setRuntimeOverrides(
                new RuntimeOverrideSettings(
                        processMatchingMode,
                        ipv6Enabled,
                        logLevel,
                        adaptiveTcpConcurrent,
                        webViewXhttp,
                        lanWebUiPublic
                ),
                callback
        ));
    }

    void networkSelectionCatalog(OperationCallback callback) {
        executor.execute(() -> {
            MihomoRuntime current = lifecycle.runtime();
            if (current == null) {
                complete(callback, false, "请先启动 VPN，再设置网络节点");
                return;
            }
            try {
                complete(callback, true, selectorSession.catalog(current));
            } catch (IOException | JSONException exception) {
                complete(callback, false, Exceptions.usefulMessage(exception));
            }
        });
    }

    void setNetworkSelection(
            String identity,
            String group,
            String target,
            OperationCallback callback
    ) {
        executor.execute(() -> {
            MihomoRuntime current = lifecycle.runtime();
            if (current == null) {
                complete(callback, false, "请先启动 VPN，再设置网络节点");
                return;
            }
            try {
                complete(callback, true, selectorSession.select(
                        current,
                        identity,
                        group,
                        target
                ));
            } catch (IOException exception) {
                complete(callback, false, Exceptions.usefulMessage(exception));
            }
        });
    }

    private void startInternal(AndroidVpnService requestedService) {
        if (requestedService == null) {
            return;
        }
        RuntimeState current = stateBus.snapshot().state();
        if (lifecycle.ownsService(requestedService)
                && (current == RuntimeState.STARTING || current == RuntimeState.RUNNING)) {
            stateBus.publish(stateBus.snapshot());
            return;
        }

        cleanupAll();
        try {
            Ipv6EnvironmentMonitor.State underlyingNetworkState = networkReconciler.start();
            publish(RuntimeState.STARTING, "正在启动同进程 mihomo JNI TUN…");
            RuntimeStartRequest request = runtimeStartRequest();
            String detail = lifecycle.start(
                    requestedService,
                    request.settings(),
                    underlyingNetworkState,
                    networkReconciler::cancelSelectionRestoration
            ) + request.detailSuffix();
            publish(RuntimeState.RUNNING, detail);
        } catch (Exception exception) {
            if (exception instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
            failActiveService("VPN 启动失败：" + Exceptions.usefulMessage(exception));
        }
    }

    @Override
    public String start() throws IOException, InterruptedException {
        RuntimeStartRequest request = runtimeStartRequest();
        return lifecycle.restart(
                request.settings(),
                networkReconciler.currentState(),
                networkReconciler::cancelSelectionRestoration
        ) + request.detailSuffix();
    }

    private RuntimeStartRequest runtimeStartRequest() {
        RuntimeOverrideSettings settings = overrideStore.settings();
        boolean localNetworkFallback = settings.lanWebUiPublic()
                && !hasLocalNetworkAccess();
        if (localNetworkFallback) {
            settings = new RuntimeOverrideSettings(
                    settings.processMatchingMode(),
                    settings.ipv6Enabled(),
                    settings.logLevel(),
                    settings.adaptiveTcpConcurrent(),
                    settings.webViewXhttp(),
                    false
            );
            overrideStore.setSettings(settings);
        }
        return new RuntimeStartRequest(
                settings,
                localNetworkFallback
                        ? " · 局域网权限未授权，WebUI 已恢复本机"
                        : ""
        );
    }

    @Override
    public boolean active() {
        return lifecycle.hasActiveService();
    }

    @Override
    public void setIdleEffectiveState(
            RuntimeOverrideSettings settings,
            Ipv6EnvironmentMonitor.State networkState
    ) {
        lifecycle.setIdleEffectiveState(settings, networkState);
    }

    @Override
    public boolean effectiveTcpConcurrent() {
        return lifecycle.effectiveTcpConcurrent();
    }

    @Override
    public int controllerPort() {
        MihomoRuntime current = lifecycle.runtime();
        return current == null ? 0 : current.controllerPort();
    }

    @Override
    public boolean hasLocalNetworkAccess() {
        return Build.VERSION.SDK_INT < ANDROID_17_API
                || context.checkSelfPermission(Manifest.permission.ACCESS_LOCAL_NETWORK)
                == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public Ipv6EnvironmentMonitor.State refreshUnderlyingNetworkState() {
        return networkReconciler.refreshState();
    }

    @Override
    public void failActiveService(String message) {
        AndroidVpnService failedService = lifecycle.service();
        cleanupAll();
        publish(RuntimeState.FAILED, message);
        if (failedService != null) {
            mainHandler.post(() -> failedService.onCoordinatorFailure(message));
        }
    }

    private void cleanupAll() {
        networkReconciler.stop();
        lifecycle.stop();
    }

    @Override
    public void submit(Runnable operation) {
        executor.execute(operation);
    }

    @Override
    public RuntimeSnapshot snapshot() {
        return stateBus.snapshot();
    }

    @Override
    public void publish(RuntimeSnapshot snapshot) {
        stateBus.publish(snapshot);
    }

    @Override
    public void publish(RuntimeState state, String detail) {
        MihomoRuntime current = lifecycle.runtime();
        boolean running = state == RuntimeState.RUNNING && current != null;
        stateBus.publish(new RuntimeSnapshot(
                state,
                detail == null ? "" : detail,
                running ? current.dashboardUrl() : "",
                running ? current.controllerPort() : 0
        ));
    }

    @Override
    public void complete(OperationCallback callback, boolean success, String message) {
        if (callback != null) {
            mainHandler.post(() -> callback.onComplete(success, message == null ? "" : message));
        }
    }

    private void postCompletion(Runnable completion) {
        if (completion != null) {
            mainHandler.post(completion);
        }
    }
}

File: app/src/main/java/io/github/qwqgong/androidcyaml/runtime/RuntimeCoordinator.kt
package io.github.qwqgong.androidcyaml

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import org.json.JSONException
import java.io.IOException
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

/** Serializes every transition that can replace the TUN or embedded mihomo runtime. */
class RuntimeCoordinator private constructor(context: Context) :
    RuntimeConfigTransactions.Lifecycle,
    RuntimeConfigTransactions.Host,
    NetworkReconciler.Host {

    fun interface OperationCallback {
        fun onComplete(success: Boolean, detail: String)
    }

    private data class RuntimeStartRequest(
        val settings: RuntimeOverrideSettings,
        val detailSuffix: String,
    )

    private val context: Context = context.applicationContext
    private val mainHandler = Handler(Looper.getMainLooper())
    private val executor = Executors.newSingleThreadExecutor()
    private val stateBus = RuntimeStateBus()
    private val fileStore = MihomoFileStore(this.context)
    private val configInstaller = ConfigInstaller(this.context, fileStore)
    private val overrideStore = RuntimeOverrideStore(this.context)
    private val selectorSession: SelectorSession
    private val lifecycle: RuntimeLifecycle
    private val networkReconciler: NetworkReconciler
    private val configTransactions: RuntimeConfigTransactions

    init {
        val networkMonitor = Ipv6EnvironmentMonitor(this.context)
        selectorSession = SelectorSession(NetworkSelectionStore(this.context), networkMonitor)
        lifecycle = RuntimeLifecycle(fileStore, selectorSession)
        networkReconciler = NetworkReconciler(
            networkMonitor,
            overrideStore,
            lifecycle,
            selectorSession,
            this,
        )
        configTransactions = RuntimeConfigTransactions(
            configInstaller,
            overrideStore,
            this,
            this,
        )
    }

    fun runtimeOverrideSettings(): RuntimeOverrideSettings = overrideStore.settings()

    override fun effectiveIpv6Enabled(): Boolean = lifecycle.effectiveIpv6Enabled

    fun addListener(listener: RuntimeStateBus.Listener) {
        stateBus.addListener(listener)
    }

    fun removeListener(listener: RuntimeStateBus.Listener) {
        stateBus.removeListener(listener)
    }

    fun start(requestedService: AndroidVpnService) {
        executor.execute { startInternal(requestedService) }
    }

    fun stop(requestedService: AndroidVpnService?, completion: Runnable?) {
        executor.execute {
            if (lifecycle.service() != null &&
                requestedService != null &&
                !lifecycle.ownsService(requestedService)
            ) {
                postCompletion(completion)
                return@execute
            }
            if (stateBus.snapshot().state != RuntimeState.STOPPED) {
                publish(RuntimeState.STOPPING, "正在停止 mihomo JNI TUN…")
            }
            cleanupAll()
            publish(RuntimeState.STOPPED, "VPN 未连接")
            postCompletion(completion)
        }
    }

    fun restart(callback: OperationCallback?) {
        executor.execute {
            if (!active()) {
                complete(callback, false, "VPN 未启动")
                return@execute
            }
            try {
                publish(RuntimeState.STARTING, "正在重启 mihomo JNI 核心…")
                publish(RuntimeState.RUNNING, start())
                complete(callback, true, "mihomo 已重启")
            } catch (exception: Exception) {
                val message = Exceptions.usefulMessage(exception)
                failActiveService("mihomo 重启失败：$message")
                complete(callback, false, message)
            }
        }
    }

    fun importConfig(source: Uri?, callback: OperationCallback?) {
        executor.execute { configTransactions.importConfig(source, callback) }
    }

    fun setRuntimeOverrides(
        processMatchingModeValue: String?,
        ipv6Enabled: Boolean,
        logLevelValue: String?,
        adaptiveTcpConcurrent: Boolean,
        webViewXhttp: Boolean,
        lanWebUiPublic: Boolean,
        callback: OperationCallback?,
    ) {
        val logLevel: RuntimeLogLevel
        val processMatchingMode: ProcessMatchingMode
        try {
            logLevel = RuntimeLogLevel.fromWireValue(logLevelValue)
            processMatchingMode = ProcessMatchingMode.fromWireValue(processMatchingModeValue)
        } catch (exception: IllegalArgumentException) {
            complete(callback, false, Exceptions.usefulMessage(exception))
            return
        }
        executor.execute {
            configTransactions.setRuntimeOverrides(
                RuntimeOverrideSettings(
                    processMatchingMode,
                    ipv6Enabled,
                    logLevel,
                    adaptiveTcpConcurrent,
                    webViewXhttp,
                    lanWebUiPublic,
                ),
                callback,
            )
        }
    }

    fun networkSelectionCatalog(callback: OperationCallback?) {
        executor.execute {
            val current = lifecycle.runtime()
            if (current == null) {
                complete(callback, false, "请先启动 VPN，再设置网络节点")
                return@execute
            }
            try {
                complete(callback, true, selectorSession.catalog(current))
            } catch (exception: IOException) {
                complete(callback, false, Exceptions.usefulMessage(exception))
            } catch (exception: JSONException) {
                complete(callback, false, Exceptions.usefulMessage(exception))
            }
        }
    }

    fun setNetworkSelection(
        identity: String?,
        group: String?,
        target: String?,
        callback: OperationCallback?,
    ) {
        executor.execute {
            val current = lifecycle.runtime()
            if (current == null) {
                complete(callback, false, "请先启动 VPN，再设置网络节点")
                return@execute
            }
            try {
                complete(callback, true, selectorSession.select(current, identity, group, target))
            } catch (exception: IOException) {
                complete(callback, false, Exceptions.usefulMessage(exception))
            }
        }
    }

    private fun startInternal(requestedService: AndroidVpnService) {
        val current = stateBus.snapshot().state
        if (lifecycle.ownsService(requestedService) &&
            (current == RuntimeState.STARTING || current == RuntimeState.RUNNING)
        ) {
            stateBus.publish(stateBus.snapshot())
            return
        }

        cleanupAll()
        try {
            val underlyingNetworkState = networkReconciler.start()
            publish(RuntimeState.STARTING, "正在启动同进程 mihomo JNI TUN…")
            val request = runtimeStartRequest()
            val detail = lifecycle.start(
                requestedService,
                request.settings,
                underlyingNetworkState,
                networkReconciler::cancelSelectionRestoration,
            ) + request.detailSuffix
            publish(RuntimeState.RUNNING, detail)
        } catch (exception: Exception) {
            if (exception is InterruptedException) {
                Thread.currentThread().interrupt()
            }
            failActiveService("VPN 启动失败：" + Exceptions.usefulMessage(exception))
        }
    }

    override fun start(): String {
        val request = runtimeStartRequest()
        return lifecycle.restart(
            request.settings,
            networkReconciler.currentState(),
            networkReconciler::cancelSelectionRestoration,
        ) + request.detailSuffix
    }

    private fun runtimeStartRequest(): RuntimeStartRequest {
        var settings = overrideStore.settings()
        val localNetworkFallback = settings.lanWebUiPublic && !hasLocalNetworkAccess()
        if (localNetworkFallback) {
            settings = settings.copy(lanWebUiPublic = false)
            overrideStore.setSettings(settings)
        }
        return RuntimeStartRequest(
            settings,
            if (localNetworkFallback) " · 局域网权限未授权，WebUI 已恢复本机" else "",
        )
    }

    override fun active(): Boolean = lifecycle.hasActiveService()

    override fun setIdleEffectiveState(
        settings: RuntimeOverrideSettings,
        networkState: Ipv6EnvironmentMonitor.State,
    ) {
        lifecycle.setIdleEffectiveState(settings, networkState)
    }

    override fun effectiveTcpConcurrent(): Boolean = lifecycle.effectiveTcpConcurrent

    override fun controllerPort(): Int = lifecycle.runtime()?.controllerPort() ?: 0

    override fun hasLocalNetworkAccess(): Boolean =
        Build.VERSION.SDK_INT < ANDROID_17_API ||
            context.checkSelfPermission(Manifest.permission.ACCESS_LOCAL_NETWORK) ==
            PackageManager.PERMISSION_GRANTED

    override fun refreshUnderlyingNetworkState(): Ipv6EnvironmentMonitor.State =
        networkReconciler.refreshState()

    override fun failActiveService(message: String) {
        val failedService = lifecycle.service()
        cleanupAll()
        publish(RuntimeState.FAILED, message)
        if (failedService != null) {
            mainHandler.post { failedService.onCoordinatorFailure(message) }
        }
    }

    private fun cleanupAll() {
        networkReconciler.stop()
        lifecycle.stop()
    }

    override fun submit(operation: Runnable) {
        executor.execute(operation)
    }

    override fun snapshot(): RuntimeSnapshot = stateBus.snapshot()

    override fun publish(snapshot: RuntimeSnapshot) {
        stateBus.publish(snapshot)
    }

    override fun publish(state: RuntimeState, detail: String) {
        val current = lifecycle.runtime()
        val running = state == RuntimeState.RUNNING && current != null
        stateBus.publish(
            RuntimeSnapshot(
                state,
                detail,
                if (running) current!!.dashboardUrl() else "",
                if (running) current!!.controllerPort() else 0,
            ),
        )
    }

    override fun complete(callback: OperationCallback?, success: Boolean, detail: String) {
        if (callback != null) {
            mainHandler.post { callback.onComplete(success, detail) }
        }
    }

    private fun postCompletion(completion: Runnable?) {
        if (completion != null) {
            mainHandler.post(completion)
        }
    }

    companion object {
        private const val TAG = "AndroidCyaml/Coordinator"
        private const val ANDROID_17_API = 37

        @Volatile
        private var instance: RuntimeCoordinator? = null

        fun getInstance(context: Context): RuntimeCoordinator {
            instance?.let { return it }
            synchronized(RuntimeCoordinator::class.java) {
                val existing = instance
                if (existing != null) {
                    return existing
                }
                val created = RuntimeCoordinator(context)
                instance = created
                return created
            }
        }

        fun trimMemoryCachesIfCreated(): Int {
            // Do not initialize the large native core merely because Android asks
            // an otherwise idle service process to trim memory.
            val current = instance?.lifecycle?.runtime()
            return current?.trimRebuildableCaches() ?: 0
        }

        fun persistStateForMemoryKill(): Boolean {
            val local = instance
            if (local == null || local.lifecycle.runtime() == null) {
                return true
            }
            val checkpoint = local.executor.submit<Boolean> {
                local.selectorSession.checkpoint(local.lifecycle.runtime())
            }
            return try {
                checkpoint.get(2L, TimeUnit.SECONDS)
            } catch (exception: InterruptedException) {
                Thread.currentThread().interrupt()
                false
            } catch (exception: ExecutionException) {
                Log.w(TAG, "Unable to persist selector choices before memory kill", exception)
                false
            } catch (exception: TimeoutException) {
                Log.w(TAG, "Unable to persist selector choices before memory kill", exception)
                false
            }
        }
    }
}
